### PR TITLE
[4] feat(balances): update balances for priority tokens

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Fixes: #3430, #3429, #3426, #3423, #3428

Since the regular balances+allowances updating polling happens only every ~30 seconds, it might be annoying if you order was just filled and you don't see balances updater. Or you just approved your token spending and still see "Approve token" state.

To fix that, I've created `PriorityTokensUpdater` that is almost the same as `BalancesAndAllowancesUpdater`.
The only difference is that the priority updater updates only assets from the current trade state + from pending orders. And it does it quite often - every ~10 seconds.

  # To Test

1. Create an order
2. Wait until it is filled
- [ ] traded assets balances should be updated within ~10 seconds
3. Approve some token
- [ ] the "Approve token" state should disappear within ~10 seconds

### CODE REVIEW

File: libs/balances-and-allowances/src/hooks/usePersistBalancesAndAllowances.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+export interface PersistBalancesAndAllowancesParams {
+  account: string | undefined
+  chainId: SupportedChainId
+  tokenAddresses: string[]
+  balancesSwrConfig: SWRConfiguration
+  allowancesSwrConfig: SWRConfiguration
+  setLoadingState?: boolean
+}
+
+export function usePersistBalancesAndAllowances(params: PersistBalancesAndAllowancesParams) {
```
Comment (shoom3301) [inline libs/balances-and-allowances/src/hooks/usePersistBalancesAndAllowances.ts:27]: The code is actually just extracted from `BalancesAndAllowancesUpdater`

Comment (anxolin) [reply libs/balances-and-allowances/src/hooks/usePersistBalancesAndAllowances.ts:27]: Makes sense that you refactored this logic so you can use it in the 2 updaters. I wonder if we should even have a sigle updater, and just pass the refresh (so any lib can tell this updater the refresh interval). 

If we do that, we might want to extract the native into its own updater, which sounds like a good idea, as its completly separated logic and clients of this lib could use it in isolation.

**Review by anxolin — APPROVED** [review]
Looks great! And is enough to get the problem solved!

One small thing though, is that, the selected token and any other tokens in the prioty queue will be queried twice. Also, it's a bit rigid in case we want finer control.

Anyways, no need to address these things, lets not overoptimise. This is already a great improvement. Let's see what are the new bottlenecks and potential improvements instead.

Really good work these 4 PRs. Thanks!!!

**Comment by anxolin** [comment]
I think we are showing 0 balance for not-connected wallets when we don't know.

image


In prod doesn't show the balance: 
image

**Comment by anxolin** [comment]
@shoom3301  Do you think is possible to trigger eagerly the refresh of a token in some situations?

For example:
- A Swap just happened: Update FROM_TOKEN, TO_TOKEN
- Wrap Ether or Unwrap: Native, WETH
- Eth Flow: Native, TO_TOKEN

**Comment by shoom3301** [comment]
>I think we are showing 0 balance for not-connected wallets when we don't know.

@anxolin nice catch! Thanks! Fixed

**Comment by shoom3301** [comment]
>Do you think is possible to trigger eagerly the refresh of a token in some situations?

@anxolin yeah, this is the next step and it would be ideal. Do you mind if I implement this in the next PR?

**Comment by anxolin** [comment]
> @anxolin yeah, this is the next step and it would be ideal. Do you mind if I implement this in the next PR?

Yes, next PR is more than enough

**Comment by elena-zh** [comment]
Looks like it takes a bit more than 10 sec to update balances on the Tokens overview page (up to 35 in my case).
not updated (10 sec)
The same with 'Approve' info: it shows a loader when a token has been approved, then it is changed to 'Approve' instead of the loader, and in about 30 sec it is changed to 'Approve'
image

**Comment by elena-zh** [comment]
Reported #3424 and #3423 for the issues reported above

**Comment by shoom3301** [comment]
@elena-zh thank you!
Will work on those issues in the next PRs, since it's nothing new (reproducible on PROD as well)

File: apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrderRow/index.tsx [hunk 2]
```diff
@@ -175,8 +175,14 @@ export function OrderRow({
 
   const showCancellationModal = orderActions.getShowCancellationModal(order)
 
+  /**
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/modules/ordersTable/pure/OrdersTableContainer/OrderRow/index.tsx:178]: @alfetopito please, take a look on the fix, I'm not sure if it's correct

File: libs/multicall/src/multiCall.ts
Comment (alfetopito) [inline libs/multicall/src/multiCall.ts:1]: @shoom3301 is it intentional to have `multicall` and `multiCall`?

image

**Pull request merged**